python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py

```diff
@@ -146,16 +159,198 @@ def get_last_updated_timestamp(record: Optional[EventLogRecord]) -> Optional[flo
+                context.job_def.asset_layer.asset_graph.get(asset_key).assets_def
+            )
+        current_timestamp = pendulum.now("UTC").timestamp()
+
+        # Explicit call to partitions def here will be replaced with AssetSlice reference once it's available.
+        partitions_def = cast(
+            Optional[TimeWindowPartitionsDefinition],
+            context.job_def.asset_layer.asset_graph.get(asset_key).partitions_def,
+        )
+        check.invariant(
+            partitions_def is None or isinstance(partitions_def, TimeWindowPartitionsDefinition),
+            "Expected partitions_def to be time-windowed.",
+        )
+
+        last_update_time_lower_bound = get_last_update_time_lower_bound(
+            deadline_cron=deadline_cron,
+            freshness_cron_timezone=freshness_cron_timezone,
+            current_timestamp=current_timestamp,
+            lower_bound_delta=lower_bound_delta,
+        )
+        expected_partition_key = get_latest_complete_partition_key(
+            deadline_cron=deadline_cron,
+            current_timestamp=current_timestamp,
+            freshness_cron_timezone=freshness_cron_timezone,
+            partitions_def=partitions_def,
+        )
+        latest_record = retrieve_latest_record(
+            instance=context.instance, asset_key=asset_key, partition_key=expected_partition_key
+        )
+        update_timestamp = get_last_updated_timestamp(latest_record)
+        passed = (
+            update_timestamp is not None
+            and update_timestamp >= last_update_time_lower_bound.timestamp()
+        )
+
+        return AssetCheckResult(
+            passed=passed,
+            description=get_description_for_freshness_check(
+                passed,
+                update_timestamp,
+                last_update_time_lower_bound,
+                current_timestamp,
+                expected_partition_key,
+            ),
+            severity=severity,
+        )
+
+    return the_check
+
+
+def get_description_for_freshness_check(
+    passed: bool,
+    update_timestamp: Optional[float],
+    last_update_time_lower_bound: datetime.datetime,
+    current_timestamp: float,
+    expected_partition_key: Optional[str],
+) -> str:
+    check.invariant(
+        (passed and update_timestamp is not None) or not passed,
+        "Should not be possible for check to pass without a record.",
+    )
+    update_time_delta_str = (
+        pendulum.duration(seconds=current_timestamp - update_timestamp).in_words()
+        if update_timestamp
+        else None
+    )
+    last_update_time_lower_bound_delta_str = pendulum.duration(
+        seconds=current_timestamp - last_update_time_lower_bound.timestamp()
+    ).in_words()
+    return (
+        f"Partition {expected_partition_key} is fresh. Expected a record for the partition within the last {last_update_time_lower_bound_delta_str}, and found one last updated {update_time_delta_str} ago."
```

Comment:
Per [this earlier comment](https://github.com/dagster-io/dagster/pull/20722/files#r1541292406), I think we should avoid talking about "records" here.